### PR TITLE
Add Vagrantfile and provisioning script

### PR DESCRIPTION
**Pull request by iandees**
Adds a Vagrantfile to make testing setup easier. The provisioning script follows the installation instructions here https://github.com/tilezen/vector-datasource/wiki/Mapzen-Vector-Tile-Service

### CODE REVIEW

File: data/Makefile-import-data.jinja2 [hunk 1]
```diff
@@ -17,7 +17,7 @@ import-shapefiles.sh: {{ tgt_shapefile_shps }}
 shapefiles: shapefiles.tar.gz
 
 shapefiles.tar.gz:
-	wget 'http://s3.amazonaws.com/{{ bucket }}/{{ datestamp }}/shapefiles.tar.gz'
+	wget --quiet 'http://s3.amazonaws.com/{{ bucket }}/{{ datestamp }}/shapefiles.tar.gz'
```
Comment (nvkelso) [inline data/Makefile-import-data.jinja2]: Add print out to say we're doing **something**, and add comment saying the `--quiet` is to be friendlier with Vagrant.

File: data/Makefile-prepare-data.jinja2 [hunk 2]
```diff
@@ -13,7 +13,7 @@ download: {{ src_shapefile_zips }}
 
 {% for shapefile in shapefiles %}
 {{ shapefile.src_zip }}:
-	wget '{{ shapefile.url }}' -O {{ shapefile.src_zip }}
+	wget --quiet '{{ shapefile.url }}' -O {{ shapefile.src_zip }}
```
Comment (nvkelso) [inline data/Makefile-prepare-data.jinja2]: Add print out to say we're doing **something**, and add comment saying the `--quiet` is to be friendlier with Vagrant.

**Comment by iandees** [comment]
@nvkelso I added a comment and echo to the Makefile to make it clearer.

File: Vagrantfile [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    sudo -u postgres psql -d osm -c 'CREATE EXTENSION postgis; CREATE EXTENSION hstore;'
+
+    export PGPASSWORD=osmpassword
+
+    mkdir -p /opt/data /opt/src
+    chown vagrant:vagrant -R /opt
+    su vagrant
+
+    # Download and import OSM data
+    wget -q -P /opt/data/ https://s3.amazonaws.com/metro-extracts.mapzen.com/new-york_new-york.osm.pbf
```
Comment (nvkelso) [inline Vagrantfile]: Would echo and quite option here and for neighbourhoods below be useful?

Comment (iandees) [inline Vagrantfile]: Yep. The `-q` is equivalent to `--quiet`. I'll switch to the longer version to make it clearer.

**Comment by iandees** [comment]
This needs a bit of doc:
- [ ] Describe what the Vagrantfile is for
- [ ] Mention that we expect `tilequeue` and `tileserver` to be next to `vector-datasource` on disk (in the same directory)

Also:
- [ ] Merge in master
- [ ] @nvkelso to try running it

**Comment by nvkelso** [comment]
I hit a snag in running `vagrant up`:

```
==> default: Downloading Who's on First neighbourhoods data
...
==> default: New python executable in env/bin/python
==> default: Installing setuptools, pip...
==> default: done.
==> default: ++ source env/bin/activate
==> default: +++ deactivate nondestructive
==> default: +++ unset pydoc
==> default: +++ '[' -n '' ']'
==> default: +++ '[' -n '' ']'
==> default: +++ '[' -n /bin/bash -o -n '' ']'
==> default: +++ hash -r
==> default: +++ '[' -n '' ']'
==> default: +++ unset VIRTUAL_ENV
==> default: +++ '[' '!' nondestructive = nondestructive ']'
==> default: +++ VIRTUAL_ENV=/opt/src/tileserver/env
==> default: +++ export VIRTUAL_ENV
==> default: +++ _OLD_VIRTUAL_PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
==> default: +++ PATH=/opt/src/tileserver/env/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
==> default: +++ export PATH
==> default: +++ '[' -n '' ']'
==> default: +++ '[' -z '' ']'
==> default: +++ _OLD_VIRTUAL_PS1='(env)'
==> default: +++ '[' x '!=' x ']'
==> default: ++++ basename /opt/src/tileserver/env
==> default: +++ '[' env = __ ']'
==> default: ++++ basename /opt/src/tileserver/env
==> default: +++ PS1='(env)(env)'
==> default: +++ export PS1
==> default: +++ alias 'pydoc=python -m pydoc'
==> default: +++ '[' -n /bin/bash -o -n '' ']'
==> default: +++ hash -r
==> default: ++ pip -q install -U -r requirements.txt
==> default:   Could not find a tag or branch 'aad32b8d5cd8848c50fbef92c73697a93cf182ba', assuming commit.
==> default: ++ python setup.py -q develop
==> default: /opt/src/tileserver/env/local/lib/python2.7/site-packages/setuptools/dist.py:331: UserWarning: Normalizing '0.8.0-pre2' to '0.8.0rc2'
==> default:   normalized_version,
==> default: error: [Errno 1] Operation not permitted: 'tileserver.egg-info/requires.txt'
The SSH command responded with a non-zero exit status. Vagrant
assumes that this means the command failed. The output for this command
should be in the log above. Please read the output to determine what
went wrong.
```

**Comment by iandees** [comment]
@rmarianski any ideas here? This looks to be a packaging problem.

**Comment by nvkelso** [comment]
The tag over there is https://github.com/tilezen/tileserver/releases/tag/v0.8.0-pre2 (not additional `v` prefix), is that the problem?

Additionally, when I type `git tag` I do see `v0.8.0-pre2` listed locally in my tileserver checkout.

**Comment by iandees** [comment]
FWIW, I built/ran this on master for all three repos.

**Comment by nvkelso** [comment]
Same here:

- tileserver <master>
- tilequeue <master>
- vector-datasource <add_vagrantfile_and_provisioning> which was merged
against master

**Comment by rmarianski** [comment]
Could it be a permissions issue?

**Comment by iandees** [comment]
On the vagrant side all files shared with the host are owned and read/writeable by the vagrant user.

I guess I'll try reproducing this later tonight on my machine.

**Comment by zerebubuth** [comment]
I ran `vagrant up` without problems. I notice that the virtualenv is in `$CWD/env` when running vagrant, which doesn't cause me a problem as mine is in a directory elsewhere on the host. I wonder if it would cause an issue for anyone already keeping their host-side virtualenv in the same directory?

**Comment by nvkelso** [comment]
My current state is an env dir in both vector-datasource and tileserver
(but not in tilequeue). Removing those and trying again...

**Comment by nvkelso** [comment]
This is probably me not being familiar with Vagrant, but since not everyone is:

When I run 
`vagrant provision`
(after it told me I needed to do that when `vagrant up` fails)

I get error:
`==> default: ERROR:  role "osm" already exists`

Full transcript:

```
$ vagrant up
Bringing machine 'default' up with 'virtualbox' provider...
==> default: Checking if box 'ubuntu/trusty64' is up to date...
==> default: A newer version of the box 'ubuntu/trusty64' is available! You currently
==> default: have version '20160913.0.0'. The latest is version '20160919.0.0'. Run
==> default: `vagrant box update` to update.
==> default: Machine already provisioned. Run `vagrant provision` or use the `--provision`
==> default: flag to force provisioning. Provisioners marked to run always will still run.
nvkelso:vector-datasource nvkelso$ vagrant provision
==> default: Running provisioner: shell...
    default: Running: inline script
==> default: stdin: is not a tty
==> default: ++ set -e
==> default: ++ debconf-set-selections
==> default: ++ echo 'debconf debconf/frontend select Noninteractive'
==> default: ++ export DEBIAN_FRONTEND=noninteractive
==> default: ++ DEBIAN_FRONTEND=noninteractive
==> default: ++ apt-add-repository -y ppa:tilezen
==> default: gpg: 
==> default: keyring `/tmp/tmpx7slkp49/secring.gpg' created
==> default: gpg: 
==> default: keyring `/tmp/tmpx7slkp49/pubring.gpg' created
==> default: gpg: 
==> default: requesting key 9C946416 from hkp server keyserver.ubuntu.com
==> default: gpg: 
==> default: /tmp/tmpx7slkp49/trustdb.gpg: trustdb created
==> default: gpg: 
==> default: key 9C946416: public key "Launchpad PPA for Tilezen" imported
==> default: gpg: 
==> default: Total number processed: 1
==> default: gpg: 
==> default:               imported: 1
==> default:   (RSA: 1)
==> default: OK
==> default: ++ apt-get -qq update
==> default: ++ apt-get -qq install -y git unzip postgresql postgresql-contrib postgis postgresql-9.3-postgis-2.1 build-essential autoconf libtool pkg-config python-dev python-virtualenv libgeos-dev libpq-dev python-pip python-pil libmapnik2.2 libmapnik-dev mapnik-utils python-mapnik osm2pgsql
==> default: ++ sudo -u postgres psql -c 'CREATE ROLE osm WITH NOSUPERUSER LOGIN UNENCRYPTED PASSWORD '\''osmpassword'\'';'
==> default: ERROR:  role "osm" already exists
The SSH command responded with a non-zero exit status. Vagrant
assumes that this means the command failed. The output for this command
should be in the log above. Please read the output to determine what
```

**Comment by nvkelso** [comment]
Reading up on the Vagrant docs says to `vagrant destroy` and then `vagrant up` again, giving that a try...

**Comment by nvkelso** [comment]
Got a little farther along, but not much:

```
==> default: Downloading Who's on First neighbourhoods data
==> default: ++ wget --quiet -P /opt/data/ https://s3.amazonaws.com/mapzen-tiles-assets/wof/dev/wof_neighbourhoods.pgdump
==> default: ++ pg_restore --clean -d osm -U osm -h localhost -O /opt/data/wof_neighbourhoods.pgdump
==> default: ++ cd /opt/src/tileserver
==> default: ++ virtualenv env
==> default: New python executable in env/bin/python
==> default: Installing setuptools, pip...
==> default: done.
==> default: ++ source env/bin/activate
==> default: +++ deactivate nondestructive
==> default: +++ unset pydoc
==> default: +++ '[' -n '' ']'
==> default: +++ '[' -n '' ']'
==> default: +++ '[' -n /bin/bash -o -n '' ']'
==> default: +++ hash -r
==> default: +++ '[' -n '' ']'
==> default: +++ unset VIRTUAL_ENV
==> default: +++ '[' '!' nondestructive = nondestructive ']'
==> default: +++ VIRTUAL_ENV=/opt/src/tileserver/env
==> default: +++ export VIRTUAL_ENV
==> default: +++ _OLD_VIRTUAL_PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
==> default: +++ PATH=/opt/src/tileserver/env/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
==> default: +++ export PATH
==> default: +++ '[' -n '' ']'
==> default: +++ '[' -z '' ']'
==> default: +++ _OLD_VIRTUAL_PS1='(env)'
==> default: +++ '[' x '!=' x ']'
==> default: ++++ basename /opt/src/tileserver/env
==> default: +++ '[' env = __ ']'
==> default: ++++ basename /opt/src/tileserver/env
==> default: +++ PS1='(env)(env)'
==> default: +++ export PS1
==> default: +++ alias 'pydoc=python -m pydoc'
==> default: +++ '[' -n /bin/bash -o -n '' ']'
==> default: +++ hash -r
==> default: ++ pip -q install -U -r requirements.txt
==> default:   Could not find a tag or branch 'aad32b8d5cd8848c50fbef92c73697a93cf182ba', assuming commit.
==> default: ++ python setup.py -q develop
==> default: /opt/src/tileserver/env/local/lib/python2.7/site-packages/setuptools/dist.py:331: UserWarning: Normalizing '0.8.0-pre2' to '0.8.0rc2'
==> default:   normalized_version,
==> default: error: [Errno 1] Operation not permitted: 'tileserver.egg-info/requires.txt'
The SSH command responded with a non-zero exit status. Vagrant
assumes that this means the command failed. The output for this command
should be in the log above. Please read the output to determine what
went wrong.
```

**Comment by musicformellons** [comment]
Looks like it is installing master. Would be nice be able to install latest required versions for the stable release for tileserver, tilequeue and mapbox-vector-tile.

**Comment by musicformellons** [comment]
After installing git on the vagrant box the locale settings should be correct otherwise postgres will be in trouble. Here is relevant info:
Before you install postgres please check your locale first. Because it can make you error during install postgres. like this one
psql: could not connect to server: No such file or directory
    Is the server running locally and accepting
    connections on Unix domain socket "/var/pgsql_socket/.s.PGSQL.5432"?
How to check locale please run command below :
$ locale
LANG=en_US.UTF-8
LANGUAGE=en_US.UTF-8
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=en_US.UTF-8

Please make sure your server doesn't miss LANGUAGE and LC_ALL. If miss run command below to update :
export LANGUAGE="en_US.UTF-8"
echo 'LANGUAGE="en_US.UTF-8"' >> /etc/default/locale
export LC_ALL="en_US.UTF-8"
echo 'LC_ALL="en_US.UTF-8"' >> /etc/default/locale

After done with setup locale you can follow by this tutorial. but if you error when type command psql
please remove postgres and setup locale first then re-install again.
There are command remove postgrss and remove dependency Postgresql too.
 $ sudo apt-get --purge remove postgresql
 $ dpkg -l | grep postgres
ii  postgresql                            8.3.17-0ubuntu0.8.04.1           object-relational SQL database (latest versi
ii  postgresql-8.3                        8.3.9-0ubuntu8.04                object-relational SQL database, version 8.3
ii  postgresql-client                     8.3.9-0ubuntu8.04                front-end programs for PostgreSQL (latest ve
ii  postgresql-client-8.3                 8.3.9-0ubuntu8.04                front-end programs for PostgreSQL 8.3
ii  postgresql-client-common              87ubuntu2                        manager for multiple PostgreSQL client versi
ii  postgresql-common                     87ubuntu2                        PostgreSQL database-cluster manager
ii  postgresql-contrib                    8.3.9-0ubuntu8.04                additional facilities for PostgreSQL (latest
ii  postgresql-contrib-8.3                8.3.9-0ubuntu8.04                additional facilities for PostgreSQL

Then get name of list and run command below
  $ sudo apt-get --purge remove postgresql postgresql-8.3 postgresql-client....

**Comment by musicformellons** [comment]
Before building makefiles these are required:
apt-get install python-pip
pip install jinja2
apt-get install unzip

**Comment by zerebubuth** [comment]
@nvkelso did you have any luck after removing the `env/` dir?

**Comment by iandees** [comment]
I think the right thing to do here is to put the virtualenv's somewhere local to the VM guest rather than using something in the project directory. I can make that change.

**Comment by nvkelso** [comment]
+1

**Comment by iandees** [comment]
The most recent change here moved the virtualenvs to a separate directory on the VM, so it should work regardless of what virtualenvs you have set up on your Mac.

**Comment by nvkelso** [comment]
Tried this again with @iandees today...

This step pauses for a long time, seems to be flaky wifi network constrained and then breaks thru:

`==> default: ++ apt-get -qq install -y git unzip postgresql postgresql-contrib postgis postgresql-9.3-postgis-2.1 build-essential autoconf libtool pkg-config python-dev python-virtualenv libgeos-dev libpq-dev python-pip python-pil libmapnik2.2 libmapnik-dev mapnik-utils python-mapnik osm2pgsql`

**Comment by nvkelso** [comment]
New fail point is:

```
==> default: New python executable in /opt/virtualenv/tileserver/bin/python
==> default: Installing setuptools, pip...
==> default: done.
==> default: ++ source /opt/virtualenv/tileserver/bin/activate
==> default: +++ deactivate nondestructive
==> default: +++ unset pydoc
==> default: +++ '[' -n '' ']'
==> default: +++ '[' -n '' ']'
==> default: +++ '[' -n /bin/bash -o -n '' ']'
==> default: +++ hash -r
==> default: +++ '[' -n '' ']'
==> default: +++ unset VIRTUAL_ENV
==> default: +++ '[' '!' nondestructive = nondestructive ']'
==> default: +++ VIRTUAL_ENV=/opt/virtualenv/tileserver
==> default: +++ export VIRTUAL_ENV
==> default: +++ _OLD_VIRTUAL_PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
==> default: +++ PATH=/opt/virtualenv/tileserver/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
==> default: +++ export PATH
==> default: +++ '[' -n '' ']'
==> default: +++ '[' -z '' ']'
==> default: +++ _OLD_VIRTUAL_PS1='(vector-datasource)'
==> default: +++ '[' x '!=' x ']'
==> default: ++++ basename /opt/virtualenv/tileserver
==> default: +++ '[' tileserver = __ ']'
==> default: ++++ basename /opt/virtualenv/tileserver
==> default: +++ PS1='(tileserver)(vector-datasource)'
==> default: +++ export PS1
==> default: +++ alias 'pydoc=python -m pydoc'
==> default: +++ '[' -n /bin/bash -o -n '' ']'
==> default: +++ hash -r
==> default: ++ pip -q install -U -r requirements.txt
==> default:   Could not find a tag or branch 'aad32b8d5cd8848c50fbef92c73697a93cf182ba', assuming commit.
==> default: ++ python setup.py -q develop
==> default: /opt/virtualenv/tileserver/local/lib/python2.7/site-packages/setuptools/dist.py:331: UserWarning: Normalizing '0.8.0-pre2' to '0.8.0rc2'
==> default:   normalized_version,
==> default: error: [Errno 1] Operation not permitted: 'tileserver.egg-info/requires.txt'
The SSH command responded with a non-zero exit status. Vagrant
assumes that this means the command failed. The output for this command
should be in the log above. Please read the output to determine what
went wrong.
```

At the very least looks like we need to rebase this branch against master again, and make sure we're using version 1.x of related tools (shows 0.8.x above).

**Comment by nvkelso** [comment]
@zerebubuth I get the same error in https://github.com/tilezen/vector-datasource/pull/1000#issuecomment-260854225 even after removing the `env/` directory before running `vagrant up`.

**Comment by zerebubuth** [comment]
`error: [Errno 1] Operation not permitted: 'tileserver.egg-info/requires.txt'`

I've seen this kind of error before when something has been trying to execute a command like:

```
CMD=`which cmd`
$(CMD) file.txt
```

When `which cmd` fails, then `$CMD` is blank and the following line attempts to execute `file.txt`, leading to the "operation not permitted" error, since `file.txt` doesn't have executable permission.

The fun part will be finding out which command is missing. That the second argument is `requirements.txt` seems to indicate it might be something like `pip`. But the log clearly shows that `pip` is installed, and the previous step which uses it completes successfully. Could you login to the VM and run the final command `python setup.py -q develop` by hand to see if that gives any more information?

**Comment by nvkelso** [comment]
Latest thinking is I installed tileserver manually which left that egg around under root instead of my account. Removing the egg and re-running `vagrant up`.

Which renders:

```
==> default: ++ pip -q install -U -r requirements.txt
==> default:   Could not find a tag or branch 'aad32b8d5cd8848c50fbef92c73697a93cf182ba', assuming commit.
==> default: ++ python setup.py -q develop
==> default: ++ cd ../tilequeue
==> default: ++ python setup.py -q develop
==> default: ++ cd ../vector-datasource
==> default: ++ python setup.py -q develop
==> default: /opt/virtualenv/tileserver/local/lib/python2.7/site-packages/setuptools/dist.py:331: UserWarning: Normalizing 'v1.0.1' to '1.0.1'
==> default:   normalized_version,
==> default: error: [Errno 1] Operation not permitted: 'vector_datasource.egg-info/requires.txt'
```

So I'll remove all the eggs from the basket and try again.

**Comment by iandees** [comment]
Yea, basically look in your `ls -l` for tilequeue, tileserver, and vector-datasource and `rm` anything owned by `root`.

**Comment by nvkelso** [comment]
That seems to have worked! At least no errors during install now :)

I will test tile serving next...

**Comment by nvkelso** [comment]
Closing as init learning task that's been completed.

**Pull request closed without merging**